packages/babel-types/src/converters/gatherSequenceExpressions.js

```diff
@@ -66,7 +66,12 @@ export default function gatherSequenceExpressions(
       exprs.push(body);
     } else if (isEmptyStatement(node)) {
       // empty statement so ensure the last item is undefined if we're last
-      ensureLastUndefined = true;
+      // checks if node is first
+      // checks if input is `;` or `;;` which will return `()` in where
+      // the length will always be 2
+      if (nodes.indexOf(node) === 0 || nodes.length === 2) {
```

Comment:
I don't think we need to check specifically `nodes.length === 2`? i.e. `do { ;;; }` should also return `undefined`.

Reply:
yeah i added the `node.length` because `()` always threw an error regardless of the amount of `;` given. 

just to clarify what does no-op mean ? 

and do we need to explicitly mark it `false` since we're in the condition that it is `true`?

Reply:
> () always threw an error regardless of the amount of ; given.

It is syntax error if there is only empty statement in the `()`, i.e. `()`, `(;)`, `(;;)` should throw.

> just to clarify what does no-op mean ?

no-op means we do not change `ensureLastUndefined`, as you can see it is reset as `false` in the beginning of loop. We can check empty statement before validating.

Let `eLU` be `ensureLastUndefined`, we can illustrate how `eLU` is changed between EmptyStatements ant other statements.

```
;                ;;;;;                              42                  ;;;;;
|--> eLU = true  |--> eLU = (previous eLU) = true   |--> eLU = false    |--> eLU= (previous eLU) = false
```

Reply:
if `()` , `(;)` , `(;;)` should throw would it be a different issue because checking if `node` is first fails the cases just mentioned?

Reply:
Babel currently throws syntax error for `()`, `(;)`, which is expected. Generally when you are working on AST, you may assume that it is always valid combinations of AST nodes, in this case, it is `EmptyStatement` nested in `BlockStatement`. The combinations of punctuators, i.e. `(`, `)` should be cared by the parser, and by the time when you have a valid AST, it can not be a representation of the source like `()`.

Generally parentheses do not affect the semantic meaning, it is a valid combination only if there is an expression wrapped inside parentheses: https://tc39.es/ecma262/#prod-ParenthesizedExpression
```
ParenthesizedExpression[Yield, Await]:
  ( Expression[+In, ?Yield, ?Await] )
```

Reply:
right, ok the changes i just did keeps the checking if `node` is first and letting it throw if `()` , `(;)`

Reply:
I think I understand what you mean by `()` and `(;)`

In the example `{ ;; }`, the parse AST is
```
[ EmptyStatement, EmptyStatement ]
```
In the first loop, `ensureLastUndefined` is set to `true`.
In the second loop, `ensureLastUndefined` is set to `false` in the beginning of the loop body: 
https://github.com/babel/babel/blob/6812a4297d1f0ecdad2ab0a67f752d9201f65b7b/packages/babel-types/src/converters/gatherSequenceExpressions.js#L28

Since `exprs` is an empty array, `ensureLastUndefined` is `false`, `sequenceExpression(exprs)` will be printed as `()`, which will throw syntax error because sequence expression can not be empty.

We can reset `ensureLastUndefined` only if `node` is not an `EmptyStatement`.

```js
for (const node of nodes) {
  if (!isEmptyStatement(node)) {
    ensureLastUndefined = false;
  }
// ...
}
```

So this ensures that when we are processing an EmptyStatement following any statements, it becomes an no-op, which does _not_ change `ensureLastUndefined`.

If we only check `node is first of nodes && nodes.length === 2`, it can not handle `do { ;;; }`: When processing the third empty statement, `ensureLastUndefined` will still be reset to `false` and `gatherSequenceExpressions` will still generate an empty sequence expression.

Reply:
That makes so much sense, thank u!